cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -588,8 +701,58 @@ inline std::array<int64_t, 3> GetIsoCalendar(int64_t arg, Localizer&& localizer)
           static_cast<int64_t>(weekday(ymd).iso_encoding())};
 }
 
-template <typename Duration>
+template <typename Duration, typename InType>
 struct ISOCalendar {
```

Comment:
Would it be possible to avoid specialization by instead making the helper functions used work with timestamps and dates? For instance, you could modify GetInputTimezone to be `GetInputTimezone<InType>` where it would always return an empty string for `InType == Date32Type`.

Reply:
(I suppose you'd also need to change functions to templates to be able to specialize things like that...)

Reply:
Thanks @lidavidm !

I think the provided implementation (using specialization) is good enough.
Please note that this function is returning a const ref string:
`const std::string& GetInputTimezone`
So the compiler will throw an error if we want to return local empty strings.
And if I change the declaration to return a `string` (without const ref) we will be copying the timezones instead of using the refs.

Also, for the case of `ISOCalendar` with `TimestampType` we had this code:
`string timezone = GetInputTimezone`
and we had been performing a copy.
So I changed that to
`const auto& timezone = GetInputTimezone`

Reply:
We could get around that with a static string somewhere, or by returning `const std::string*`. Or the actual implementation could be factored out, and the date kernel could call it with a literal string. At the very least, I'm not super enthused about essentially copy-pasting this kernel.

Reply:
Sure, that's a good idea, I can try to have a single implementation.
Where should I put the empty static string? in which file?

Reply:
It would go in this file, though FWIW, it would probably be easiest to just factor things into functions since these are all static methods anyways.

Reply:
@lidavidm I think this last version reuse the kernel logic and it specializes only the parts related to the timestamp logic. Please let me know if if good enough.
TODO: I need to fix the some R tests (I'll tackle this tomorrow)